### DISCUSSION TITLE
How to suppress/disable "deeply nested key returned undefined" warning

### DISCUSSION THREAD

**Discussion by MattChowski** [TanStack/table#4788] [Q&A]
Hello!

This seems to be a rather simple question but no idea how to solve it. From our backend, some values can return "undefined", I do handle this case in the table so nothing breaks.

Being fully aware of this issue, I'd like to remove the warning from the console as it obstructs other messages. Is there any way to do so? I have no ways to change the values from the backend to return null instead.

I've tried to look for the issue and the only thing I found is that this used to log an error instead of a warning and decision has been made to keep the warning.

Thanks in advance!

**Comment by devthub** [comment]
DO you have any work around for this?

**Reply by MattChowski** [reply]
Nope

**Reply by KevinVandy** [reply]
You can use an `accessorFn` instead of the string `accessorKey`. In your accessorFn, you can handle the undefined checks how you want.

**Reply by danikane** [reply]
That's what I ended up with then I saw this answer. :)

**Comment by briavicenti** [comment]
This is also an issue for us. Our tables have placeholder logic to show a dash in the event that a given row value is `undefined`, and so we don't benefit from the warnings being thrown. I implemented a custom `accessorFn` that uses the Lodash `_.get` helper to retrieve a field's value with a nested key, but it would have been nice to use the React-Table's built-in functionality for this instead!

**Comment by strimbob** [comment]
bump! i also think this warning should be configurable , it is super annoying.I implemented a custom accessorFn too

**Comment by martinliptak** [comment]
What's the point of showing the warning? Just handle it the same way as null. For example, we show a list of orders along with their supplier's name. Some orders don't have a supplier so "supplier.name" is undefined. I don't find any value in the warning.

**Reply by MichaelDimmitt** [reply]
I could see this going either way. I read through this pr:
https://github.com/TanStack/table/pull/4678 and the pr before it https://github.com/TanStack/table/pull/4620

This is the line of code in question: [L105](https://github.com/TanStack/table/blob/66fbe3a74b129e700e4667d1d5e1b29dd0a7b723/packages/table-core/src/core/column.ts#L105)

Total guess, but I think the intention here was to help the developer if the nested property they filled in was an incorrect key-name.
Maybe @KevinVandy , is more filled in on the subject.

On the other hand, sometimes you get data back from the api team and they are not giving you nulls they are giving undefined mixed with normal values. So there are warnings everywhere which is annoying.

And I do not want to override the accessorFn everywhere throughout the application.

I also do not want to filter every api request to coerce all undefined's coming back to be null or empty string but that is an option and I have seen it done before in this scenario. (It is just selling it to the team I might need help with 😅)

But from the pr's I have looked at so far it looks like no one has submitted on yet to just add a config to the table to prevent these warnings. So maybe I could submit one and see what happens.
I would probably put the config in the table options. Similar to [debugCells-L99](https://github.com/TanStack/table/blob/66fbe3a74b129e700e4667d1d5e1b29dd0a7b723/packages/table-core/src/core/table.ts#L99)

We probably have 1000's of these warnings throughout the codebase I work within. smh. 😅
This is the 7th most upvoted discussion so might not be top priority.

**Reply by MichaelDimmitt** [reply]
update: the accessorFn change was not as bad as I thought it would be.
others hinted to it above but not in a simple way.

```typescript
// old - causes warning
accessorKey: 'car.color',
```

```typescript
// change to this to avoid the warning.
accessor: (originalRow) => originalRow.car?.color,
 ```
 as seen here:
 https://github.com/TanStack/table/issues/579#issuecomment-947831467
 
 note, if you do it this way you need to add the option chaining yourself for however deeply nesting you want to go.

**Reply by MichaelDimmitt** [reply]
Unfortunately, you have to have both for sorting to work.
So if you are trying to get rid of the warning caused from:
```typescript
accessorKey: 'originalRow.car.color',
```

Now you need both and your accessorFn needs to handle its own partials for each level of nesting might have undefined.
I confirmed sorting was working on my end after making these changes. I will probably submit some config driven pr sometime this week just to get the discussion started. Because this is a lot of boiler plate just to avoid a warning. And it is error prone if you miss the option chain on the accessorFn.
```typescript
accessorKey: 'originalRow.car.color',
accessorFn: (originalRow) => originalRow.car?.color,
```

**Reply by cuongdo926** [reply]
I believe this will eliminate the warning and work for sorting without needing to define both.

`accessorFn: () => "car.color"`

**Reply by KevinVandy** [reply]
Are you not using an `id` with the accessorFn? You should only need an accessorFn

**Reply by MichaelDimmitt** [reply]
@KevinVandy, 

Maybe I was wrong using accessorKey instead of id. I think it got on that path in my goal to get it working in as few changes as possible. I was not using id with the accessorFn. 

**Conclusion:**
Either id, accessorKey, or header were need for sorting to work.

**Note:**
Oddly enough, if id is not provided accessorKey will be used; and if that is not provided header will be used as the id.
Note: I was using [server-side-sorting](https://www.material-react-table.com/docs/guides/sorting#manual-server-side-sorting)

**Another Note:**
If id, accessorKey, and header are not present and only accessorFn is present there is a runtime error telling you to put the id. I did not run into this runtime error since I added an accessorKey and had a header on the column.

```javascript 
// Recommended approach. ✅
const someColumns = [{
        id: 'originalRow.car.color',
        accessorFn: (originalRow) => originalRow.car?.color,
        header: 'Car Color.',
}]
```
```javascript 
// Also works but is probably not recommended. ❌
const someColumns = [{
        accessorKey: 'originalRow.car.color',
        accessorFn: (originalRow) => originalRow.car?.color,
        header: 'Car Color',
}]
```
console logging sorting in any of the scenarios described will show the following:
```
sorting: [
    {
        "id": "originalRow.car.color",
        "desc": false
    }
]
```

**Reply by MichaelDimmitt** [reply]
@cuongdo926, simply adding accessorFn will eliminate the nested warning but sorting will not work the simplest example of this I could give you is the following column.

```javascript
const someColumns = [{
        accessorFn: (originalRow) => originalRow.car?.color,
        header: 'Car Color',
}]
```
 [server-side-sorting](https://www.material-react-table.com/docs/guides/sorting#manual-server-side-sorting) will not work in this circumstance because it does not see your sort key as { id: 'originalRow.car.color' }, rather the id passed is the header string.

**Reply by cuongdo926** [reply]
> [Unique Column IDs](https://tanstack.com/table/latest/docs/guide/column-defs#unique-column-ids)
Columns are uniquely identified with 3 strategies:
If defining an accessor column with an object key or array index, the same will be used to uniquely identify the column.
Any periods (.) in an object key will be replaced by underscores (_).
If defining an accessor column with an accessor function
The columns id property will be used to uniquely identify the column OR
If a primitive string header is supplied, that header string will be used to uniquely identify the column
🧠 An easy way to remember: If you define a column with an accessor function, either provide a string header or provide a unique id property.

Apologies for the brief comment earlier; it didn't clearly explain my solution. Here is my schema and column definition. I'm working on server-side sorting, and the id value is what the server expects. I hope the follow code helps illustrate my explanation.

```ts

// schema
export const taskSchema = z.object({
  id: z.number(),
  priority: z
    .object({
      id: z.number(),
      name: z.string(),
    })
    .nullish(),
  title: z.string()
});
export type TaskItem = z.infer<typeof taskSchema>;

// column definitions
const columnHelper = createColumnHelper<TaskItem>();
export const columns = [
  columnHelper.accessor('title', {
    header: ({ column }) => <TableColumnHeader column={column} title="Title" />,
    cell: ({ renderValue }) => renderValue(),
  }),

  // NOT WORK. No warning, but sorting doesn't work
  // columnHelper.accessor((x) => x.priority?.name, {
  //   id: 'priority.name',
  //   header: ({ column }) => <TableColumnHeader column={column} title="Priority" />,
  //   cell: ({ row }) => row.original.priority?.name,
  //   meta: {
  //     visibilityName: 'priority',
  //   },
  // }),

  // NOT WORK. Warning: "priority" in the nested key "priority.name" is undefined.
  // columnHelper.accessor('priority.name', {
  //   id: 'priority.name',
  //   header: ({ column }) => <TableColumnHeader column={column} title="Priority" />,
  //   cell: ({ row }) => row.original.priority?.name,
  //   meta: {
  //     visibilityName: 'priority',
  //   },
  // }),

  // WORKS but the accessor must match a schema key
  // columnHelper.accessor('priority', {
  //   id: 'priority.name',
  //   header: ({ column }) => <TableColumnHeader column={column} title="Priority" />,
  //   cell: ({ row }) => row.original.priority?.name,
  //   meta: {
  //     visibilityName: 'priority',
  //   },
  // }),

  // WORKS. The accessor can be any value — it doesn't have to match a schema key.
  columnHelper.accessor(() => 'priority.name_or_anything', {
    id: 'priority.name', // react-table sorting state: "sorting":[{"id":"priority.name","desc":false}]
    header: ({ column }) => <TableColumnHeader column={column} title="Priority" />,
    cell: ({ row }) => row.original.priority?.name,
    meta: {
      visibilityName: 'priority',
    },
  }),
  columnHelper.accessor('id', {
    header: ({ column }) => <TableColumnHeader column={column} title="Id" />,
    cell: ({ row }) => row.original.id,
    size: 48,
  }),
  columnHelper.display({
    id: '_actions',
    cell: ({ row }) => <DataTableRowActions row={row} />,
    enableHiding: false,
    size: 60,
  }),
];
```

The JSON payload I send to the server.
````json
{"page":1,"pageSize":10,"sort":[{"id":"priority.name","desc":true}],"filters":[]}
````
